File: answers/leetcode/Symmetric Tree/Symmetric Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        return root == null || isMirror(root.left, root.right);
    }

    private boolean isMirror(TreeNode left, TreeNode right) {
        return left == right || (left != null && right != null && left.val == right.val && isMirror(left.left, right.right) && isMirror(left.right, right.left));
    }
}

File: answers/leetcode/Symmetric Tree/Symmetric Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """

        return None == root or (None == root.left and None == root.right) or self.isMirror(root.left, root.right)

    def isMirror(self, left, right):
        return (None == left and None == right) or (None != left and None != right and left.val == right.val and self.isMirror(left.left, right.right) and self.isMirror(left.right, right.left))
